### ISSUE TITLE
Need to re-infer collection and element type based on inferred collection type

### ISSUE THREAD

**Issue by 333fred** [dotnet/roslyn#33257]
```C#
using System.Collections.Generic;
class C
{
    void M(object? o)
    {
        L(o)[0].ToString(); // Should get a warning
        foreach (var x in L(o))
        {
            x.ToString(); // Should get a warning
        }
    }

    List<T> L<T>(T t) => null!;
}
```

jcouv updated: Relates to https://github.com/dotnet/roslyn/issues/29964 (need to update indexer based on tracked type of receiver)

**Comment by 333fred** [comment]
As part of this, we also need to make sure this works as expected:
```C#
using System.Collections.Generic;
class C
{
    void M(object? o)
    {
        if (o == null) return;
        foreach (var x in L(o))
        {
            x.ToString(); // Should get a warning
        }
    }

    List<T> L<T>(T t) => null!;
}
```

**Comment by 333fred** [comment]
Another broken scenario:
```C#
class C
{
    void M(string? s)
    {
        var l1 = new[] { s };
        foreach (var x in l1)
        {
            x.ToString(); // Should warn
        }

        if (s == null) return;
        var l2 = new[] { s };
        foreach (var x in l2)
        {
            x.ToString();
        }
        
    }
}
```

**Comment by jaredpar** [comment]
The update on condition checks are covered by other bugs but should verify here.

**Comment by jcouv** [comment]
The indexer issue (needs to update based on inferred receiver type) is tracked by https://github.com/dotnet/roslyn/issues/29964 already.
I updated the title for this issue to focus on the remaining problem with `foreach`.
